docs/pwclip.1
.\" Man page generated from reStructuredText.
.
.TH "PWCLIP" "1" "Feb 19, 2020" "1.7.2" "pwclip"
.SH NAME
pwclip \- pwclip documentation
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH DESCRIPTION
.sp
pwclip is a password management tool. It\(aqs main target is having fast and
comfortable access to passwords by storing them for a variable time in the
systems clipboard (copy/paste buffer). It uses either GnuPG2 or OpenSSL
(converted to GPGSM) keys for cryptographic operations. It also works with
yubikey\(aqs challenge\-response to generate uniq HMAC\-SHA1 hashes as well.
.sp
The main feature is the pwcli/pwclip mode which provides easy access to the
~/.passcrypt file. That file is used by default as password storage. The
program is executed in gui mode when pwclip is called and in cli mode when
pwcli is called respectivly. It creates the ~/.passcrypt file which is gpg
encrypted text using either the value of GPGKEY or GPGKEYS from the users
environment as gpg encryption recipients. On startup it lookes for a
~/.pwd.yaml file which is merged with the already known passwords from the
~/.passcrypt file if there already is one. All entrys in the ~/.passcrypt file
will be overridden by entrys from the ~/.pwd.yaml file. Adding, removing or
changing password entrys is also possible using pwcli mode.
.sp
The second operating mode is for operating on yubikeys to generate uniq
responses which might be used as passwords while they can be generated by that
exact yubikey only. The first yubikey found on the system and the first slot,
configured with (HMAC\-SHA1) challenge\-response, will be used. For that function
Windows is supported (see "Install" section).
.sp
To catch user input python\(aqs Tk (tkinter) library is used to create a simple
password input window. The appropriate response is saved for only 3 seconds
by default to not have it exposed as soon as it\(aqs used. The utility also
supports the input of any integer which is then used as timer. Otherwise the
environment is searched for PWCLIPTIME and uses the value of that environment
variable as timer. The timer is used as time in which the received password
stays in the paste buffer bevore its replaced by the previously copied value.
As you may see there is an optional commet which is used as text notification
displayed on the screen if set. Therefor python3\(aqs gi notify2 is used which is
another reason for discontinuing python2 support.
.sp
I would encourage you to bind pwclip to a shortcut within your X\-Environment
to have access to your passwords at any time. On Windows\-Systems you need to
create a link for it somewhere. When editing that link you may set a keyboard
shortcut (could not find a nicer solution by now). The target for that link
then whould be "%PYTHONINSTALLDIR%\escripts\epwclip.exe".
.sp
I\(aqve been trying my best to keep the passwords from unwanted access BUT i do
!NOT GUARANTEE! that the passwords handled during runtime are safe from other
users access (especially "root" on linux systems \- help on that is very
welcome) \- please be aware of that!
.sp
Since version 1.2 openssl keys are supported. For the use with pwclip they will
be converted to gpg\-keys automaticly and gpgsm is used instead of gpg \- openssl
is not used for en/decryption to be precisely.
.sp
The ~/.pwd.yaml as well as the ~/.passcrypt file is assembled as a list of one
or more users which should be named like the username used for current system
login. You may have more than one users passwords in the passcrypt but any user
who can store passwords also is able to read password stored for other users in
that file. It is just of use for visual kategories if you have stored passwords
for different systems users sharing the same passcrypt file.
As an example: A person called "Andrew" has a username called "bob". On some
other machine he has a username called "cat" and he uses SSH for both of them.
The shared passcrypt file would be assembled like the attached YAML\-Format
section. If so providing \-u or \-A when beeing logged in as "bob" or "cat"
respectivly may be omitted. In the following example both, "bob" and "cat" have
an entry called ssh. If \-A is used in that case the entry corresponding to the
logged\-in user is prefered. If in doubt you\(aqre able to read the crypt file
using "gpg \-d ~/.passcrypt" as well (assuming the correct gpg\-key is present).
.SH YAML-FORMAT
.sp
\-\-\-
.INDENT 0.0
.TP
.B bob:
\- ssh:
.INDENT 7.0
.INDENT 3.5
\- $om3(rypt!cPass
.sp
\- DoNotForgetSomethingReminder
.UNINDENT
.UNINDENT
.sp
\- othertool:
.INDENT 7.0
.INDENT 3.5
\- 0th3r(rypt!cPass
.UNINDENT
.UNINDENT
.TP
.B cat:
\- ssh:
.INDENT 7.0
.INDENT 3.5
\- $om30ther(rypticPass
.UNINDENT
.UNINDENT
.UNINDENT
.SH SYNOPSIS
.INDENT 0.0
.INDENT 3.5
\fBpwclip\fP  [\fIoptions\fP]
.UNINDENT
.UNINDENT
.SH OPTIONS
.INDENT 0.0
.TP
.B \-\-version
show program\(aqs version number and exit
.UNINDENT
.INDENT 0.0
.TP
.B \-D, \-\-debug
debugging mode
.UNINDENT
.INDENT 0.0
.TP
.B \-A, \-\-all
switch to all users entrys ("d0n" only is default)
.UNINDENT
.INDENT 0.0
.TP
.B \-o, \-\-stdout
print password to stdout (insecure and unrecommended)
.UNINDENT
.INDENT 0.0
.TP
.B \-s, \-\-show\-passwords
show passwords when listing (replaced by "*" is default)
.UNINDENT
.INDENT 0.0
.TP
.B \-t seconds
time to wait before resetting clip (3 is default)
.UNINDENT
.INDENT 0.0
.TP
.B \-p PWD, \-\-password PWD
enter password for add/change actions (insecure & not recommended)
.UNINDENT
.INDENT 0.0
.TP
.B \-\-comment COM
enter comment for add/change actions
.UNINDENT
.INDENT 0.0
.TP
.B \-R
use remote backup given by \-\-remote\-host
.UNINDENT
.INDENT 0.0
.TP
.B \-\-remote\-host HOST
use HOST for connections
.UNINDENT
.INDENT 0.0
.TP
.B \-\-remote\-user USER
use USER for connections to HOST ("d0n" is default)
.UNINDENT
.INDENT 0.0
.TP
.B \-r "ID ...", \-\-recipients "ID ..."
one ore more gpg\-key ID(s) to use for encryption (strings seperated by spaces within "")
.UNINDENT
.INDENT 0.0
.TP
.B \-u USER, \-\-user USER
query entrys only for USER (\-A overrides, "d0n" is default)
.UNINDENT
.INDENT 0.0
.TP
.B \-x, \-\-x509
force ssl compatible gpgsm mode \- usually is autodetected (use \-\-cert & \-\-key for imports)
.UNINDENT
.INDENT 0.0
.TP
.B \-C SSL\-Certificate, \-\-cert SSL\-Certificate
one\-shot setting of SSL\-Certificate
.UNINDENT
.INDENT 0.0
.TP
.B \-K SSL\-Private\-Key, \-\-key SSL\-Private\-Key
one\-shot setting of SSL\-Private\-Key
.UNINDENT
.INDENT 0.0
.TP
.B \-\-ca SSL\-CA\-Certificate, \-\-ca\-cert SSL\-CA\-Certificate
one\-shot setting of SSL\-CA\-Certificate
.UNINDENT
.INDENT 0.0
.TP
.B \-P CRYPTFILE, \-\-passcrypt CRYPTFILE
set location of CRYPTFILE to use as password store (~/.passcrypt is default)
.UNINDENT
.INDENT 0.0
.TP
.B \-Y YAMLFILE, \-\-yaml YAMLFILE
set location of YAMLFILE to read whole sets of passwords from a yaml file (~/.pwd.yaml is default)
.UNINDENT
.INDENT 0.0
.TP
.B \-S {1,2}, \-\-slot {1,2}
set one of the two yubikey slots (only useful with \-y)
.UNINDENT
.INDENT 0.0
.TP
.B \-y [SERIAL], \-\-ykserial [SERIAL]
switch to yubikey mode and optionally set SERIAL of yubikey (autoselect serial and slot is default)
.UNINDENT
.INDENT 0.0
.TP
.B \-a ENTRY, \-\-add ENTRY
add ENTRY (password will be asked interactivly)
.UNINDENT
.INDENT 0.0
.TP
.B \-c ENTRY, \-\-change ENTRY
change ENTRY (password will be asked interactivly)
.UNINDENT
.INDENT 0.0
.TP
.B \-d ENTRY [ENTRY ...], \-\-delete ENTRY [ENTRY ...]
delete ENTRY(s) from the passcrypt list
.UNINDENT
.INDENT 0.0
.TP
.B \-l [PATTERN], \-\-list [PATTERN]
pwclip an entry matching PATTERN if given \- otherwise list all entrys
.UNINDENT
.SH USAGE
.INDENT 0.0
.INDENT 3.5
Although is was planed as GUI\-Program it\(aqs also possible to be executed from
terminals. For Windows, Linux and OSX there is an appropriate executable
packed which might be executed like the following examples will show:
.sp
\fB\-GPG\-Mode\-\fP
.sp
If there is an environment variable called GPGKEYS it will use those keys to
encrypt on changes to the password file. To list the password file you may use
the list switch followed by optional search pattern like:
.sp
\fBpwcli \-l\fP
.sp
or
.sp
\fBpwcli \-l $PATTERN\fP
.sp
as you can see the yaml format tends to be used for multiple user names to
better manage large lists. By default the current users entrys will be listed
only. To have them all listed (or searched for by the above pattern example)
use:
.sp
\fBpwcli \-A \-l $PATTERN\fP
.sp
To one\-shot convert a key/cert pair in openssl x509 format, read passwords from
passwords.yaml and list them:
.sp
\fBpwcli \-Y passwords.yaml \-\-cert ssl.crt \-\-key ssl.key \-\-ca\-cert ca.crt \-l\fP
.sp
\fB\-Yubikey\-Mode\-\fP
.sp
The YKSERIAL environment variable is used if found to select the yubikey to use
if more than one key is connected. Otherwise the first one found is chosen.
Likewise it also accepts an option:
.sp
\fBpwcli \-y $YKSERIAL\fP
.sp
To have it wait for a specific time like 60 seconds (bevore resetting the paste
buffer to the previously copied value) the PWCLIPTIME environment variable is
used or also the command accepts it as input:
.sp
\fBpwcli \-t 60 \-l somename\fP
.sp
Most of the options may be combined. For more information on possible options in
cli mode please see:
.sp
\fBpwcli \-\-help\fP
.sp
\fB\-GUI\-Modes\-\fP
.sp
For the GUI\-Mode just use one of the following commands, also accepting most of
the commandline arguments:
.sp
\fBpwclip\fP
.sp
\fBykclip\fP
.UNINDENT
.UNINDENT
.SH TROUBLESHOOT
.sp
When using the yubikey challenge\-response mode there is a bug in the usb_hid
interface. This is because of python2 => 3 transition, most likely and can be
fixed by executing the following command:
.sp
\fBsudo vi +\(aq:107s/\e(.* =\e).*/\e1 response[0]/\(aq +\(aq:wq\(aq /usr/local/lib/python3.5/dist\-packages/yubico/yubikey_4_usb_hid.py\fP
.sp
Explained:
.sp
In line 107 of the file
.sp
\fB/usr/local/lib/python3.5/dist\-packages/yubico/yubikey_4_usb_hid.py\fP
.sp
the ord() coversion of the response:
.sp
\fBr_len = ord(response[0])\fP
.sp
needs to be replaced by:
.sp
\fBr_len = response[0]\fP
.SH CREDITS
.sp
I hope that this might be somewhat of help or at least be inspiring for own
ideas. You\(aqre alway welcome to leave me a message for reviews or feature
requests as well as bug reports: <\fI\%mail@leonpelzer.de\fP>
.INDENT 0.0
.INDENT 3.5
.INDENT 0.0
.IP \(bu 2
Python3 developers for IMHO one of the best programming languages
.IP \(bu 2
stackoverflow.com for hosting endless threads of troubleshooting
.IP \(bu 2
Pyperclip for the excellent Windows & OSX clipboard code
.IP \(bu 2
Conrad Parker for xsel as linux copy/paste backend
.IP \(bu 2
Yubico (cheap & solid HW\-Security\-Modules) & python\-yubico developers
.IP \(bu 2
GNU Privacy Guard (basic kryptography) & python\-gnupg developers
.IP \(bu 2
SonicLux for testing and telling me that a final version must not be 0.3.3 :D
.UNINDENT
.UNINDENT
.UNINDENT
.SH SEE ALSO
.INDENT 0.0
.INDENT 3.5
python3(1), gpg2(1), gpgsm(1), openssl(1), ssh(1), yaml(1)
.UNINDENT
.UNINDENT
.SH AUTHOR
Leon Pelzer
.SH COPYRIGHT
2018, Leon Pelzer
.\" Generated by docutils manpage writer.
.


CHANGELOG.rst
Changelog
=========

1.2.10 (current)
---------------

Released: 2018-03-31

  * fix ykclip not choosing challenge-response slot automatically

  * changed --help for -x option to better match its sence

  * again lots of code linting (code rated at 9.52/10) disabled checks:
    bad-continuation,too-many-arguments,too-many-locals,too-many-branches


1.2.9
-----

Released: 2018-03-30

  * fixed xsel stays running in background (it's options are nasty)

  * lots of merges but it's too fuzzy to take them apart to see what has
    changed - mostly no functionallity changes though

  * replaced the xinput and xgetpasswd (each continaing the class XInput) by
    the equal called functions each using the class XInput

  * added message for failed input match or decryption

  * fixed displaying messages for input windows

  * reverted and finally fixed xsel running in background (sometimes?)
    now explicitly terminating when finished


1.2.8
-----

Released: 2018-03-29

  * fixed a few bugs implemented while linting in 1.2.5

  * made GPGSMTool parent to GPGTool for common methods

  * fixed wrong filerotate if no changes where made

  * fixed bug in class import name


1.2.7
-----

Released: 2018-03-28

  * fix for passwords are returned incorrectly

  ** hotfix release **


1.2.6
-----

Released: 2018-03-27

  * example include fixed

  * stylistic updates and code cleanup


1.2.5
-----

Released: 2018-03-27

  * lotz of linting - almost everything fixed complying to pylint3

  * renamed gpg module to gpgtool due to namespace restrictions


1.2.4
-----

Released: 2018-03-27

  * fixed filerotate not copying file modes as well

  * fixed encrypt function when setting recipients via user environment
    variables (GPGKEY/GPGKEYS) only

  * fixed creating a new password crypt file from scratch

  * minor cleanup in gpg module

  * added example files in addition to README and --help

  * fixed some obsolete info in README


1.2.3
-----

Released: 2018-03-22

  * some minor bugfixes to comply with new executor features (bytes, input)

  * fixed unintended executions when using TAB in a shell

  * some cosmetics for `pwcli --help` message

  * reimplemented the timer for remote access & sync to prevent unnecessary
    connection delays (if remote is used -R or config file)

  * another fix for executors byte2string feature - the default is to encode
    input strings and decode output strings what broke gpgsm en/decryption

  * [W] fixed scp put/get for windows (skipped instead of rising exceptions)


1.2.2
-----

Released: 2018-03-22

  * merged updates for executor subprocess forker to accept input and be able
    to return stdout as byte-string

  * removed unnecessary library iface from net

  * fixed ssh.put/get function if no scp available

  * fix for gpg socket location changed in newer releases - should also comply
    with older versions as well


1.2.1
-----

Released: 2018-03-18

  * [W] fixed some wrong path concatenations

  * another few lib fixes merged

  * fixed wrong current version in changelog


1.2.0
-----

Released: 2018-03-18

  * [L] fixed paramiko scp not working any more (replaced by subprocess call)

  * added file rotation for passcrypt file

  * cleanup of unused imports and libraries

  * removed unnecessary read/write actions on passcrypt

  * class GPGSMTool added to interact wit gpgsm for openssl compatibility


1.1.12
------

Released: 2017-12-12

  * fixed typo in system.which function changed to comply for windows

  ** hotfix release **


1.1.11
------

Released: 2017-12-11

  * [W] fix gpg.exe not found by system.which function

  * trying to fix some password input - gpg related issues

  * changed README file a bit for windows installation

  ** still some work todo for implementing gpg-key generating functionality **


1.1.10
------

Released: 2017-11-17

  * correcture on last release date :P

  * reverted which and gpg module *yet another hotfix release*


1.1.9
-----

Released: 2017-11-17

  * reverted system.user.whoami module to last commit *hotfix release*


1.1.8
-----

Released: 2017-11-16

  * fixed TypeError if password is an integer

  * fixed alot of stuff in gpg and passcrypt module for key-gen function
    to comply to gui mode as well

  * cleaned up remains of submodule merges


1.1.7
-----

Released: 2017-11-13

  * fixed key-gen dialog in cli and gui mode

  * fixed gpg-findkey function in secret-key-mode to not ask for password

  * fixed lotz of stuff in gpg wrapper for correctly collecting user input

  * merged almost all modules back into master and cleaned up pwclip branch

  * added xgetpass module and removed character hiding in xinput module

  * rearranged build environment with git-submodules


1.1.6
-----

Released: 2017-11-06

  * fixed issue where existing gpg-keys would not be recognised

  * fixed some message typos

  * continued implementing key-gen function when secret-key is missing


1.1.5
-----

Released: 2017-11-05

  * fixed date in changelog and other documentation fails from last release

  * fixed secret key listing requires password

  * still working on generating gpg-key functionality (slomo)


1.1.4
-----

Released: 2017-11-04

  * hotfix release for failed last upload


1.1.3
-----

Released: 2017-11-04

  * fixed some changelog entrys and release date of last release in changelog

  * fixed which function to return only absolute paths


1.1.2
-----

Released: 2017-11-01

  * [W] added missing wget dependency for gpg4win installation

  * [W] fix download & install gpg4win in gui mode

  * [W] fixed gpg2.exe was used in some cases (gpg2 does not work on windows)

  * added changelog entry for the last release

  * implementing key-gen dialog if no secret-keys found
    (incomplete & unapplied)


1.1.1
-----

Released: 2017-10-24

  * [W] fix for readline import not working on windows

  * made input readline compatible if on linux


1.1.0
-----

Released: 2017-10-12


pwclip.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
#
# This file is free software by d0n <d0n@janeiskla.de>
#
# You can redistribute it and/or modify it under the terms of the GNU -
# Lesser General Public License as published by the Free Software Foundation
#
# This is distributed in the hope that it will be useful somehow.
# !WITHOUT ANY WARRANTY!
#
# Without even the implied warranty of MERCHANTABILITY or FITNESS FOR
# A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
# details.
#
"""pwclip main program"""
# global & stdlib imports
try:
	from os import fork
except ImportError:
	def fork(): """fork faker function""" ;return 0

from os import environ, path, remove, name as osname

from subprocess import call

from argparse import ArgumentParser

from argcomplete import autocomplete
#from argcomplete.completers import FilesCompleter, ChoicesCompleter

from time import sleep

from yaml import load

try:
	import readline
except ImportError:
	pass

# local relative imports
from colortext import bgre, tabd, error, fatal

from system import copy, paste, xgetpass, xmsgok, xyesno, xnotify, which

# first if on windows and gpg.exe cannot be found in PATH install gpg4win
if osname == 'nt' and not which('gpg.exe'):
	if not xyesno('gpg4win is mandatory! Install it?'):
		exit(1)
	import wget
	src = 'https://files.gpg4win.org/gpg4win-latest.exe'
	trg = path.join(environ['TEMP'], 'gpg4win.exe')
	wget.download(src, out=trg)
	try:
		call(trg)
	except TimeoutError:
		exit(1)
	finally:
		remove(trg)

from secrecy import PassCrypt, ykchalres

from __pkginfo__ import version

def forkwaitclip(text, poclp, boclp, wait=3):
	"""clipboard forking, after time resetting function"""
	if fork() == 0:
		try:
			copy(text, mode='pb')
			sleep(int(wait))
		except (KeyboardInterrupt, RuntimeError):
			exit(1)
		finally:
			copy(poclp, mode='p')
			copy(boclp, mode='b')
	exit(0)

def confpars():
	"""pwclip command line opt/arg parsing function"""
	prol = 'pwclip - multi functional password manager to temporarily ' \
           'save passphrases  to your copy/paste buffers for easy and ' \
           'secure accessing your passwords'
	pars = ArgumentParser(description=prol) #add_help=False)
	pars.set_defaults(**cfgs)
	pars.add_argument(
        '--version',
        action='version', version='%(prog)s-v'+version)
	pars.add_argument(
        '-D', '--debug',
        dest='dbg', action='store_true', help='debugging mode')
	pars.add_argument(
        '-A', '--all',
        dest='aal', action='store_true',
        help='switch to all users entrys (instead of current user only)')
	pars.add_argument(
        '-o', '--stdout',
        dest='out', action='store_true',
        help='print received password to stdout (insecure & unrecommended)')
	pars.add_argument(
        '-s', '--show-passwords',
        dest='sho', action='store_true',
        help='switch to display passwords (replaced with * by default)')
	pars.add_argument(
        '-t',
        dest='time', default=3, metavar='seconds', type=int,
        help='time to wait before resetting clip (default is 3 max 3600)')

	rpars = pars.add_argument_group('remote arguments')
	rpars.add_argument(
        '-R',
        dest='rem', action='store_true',
        help='use remote backup given by --remote-host')
	rpars.add_argument(
        '--remote-host',
        dest='rehost', metavar='HOST',
        help='use HOST for connections')
	rpars.add_argument(
        '--remote-user',
        dest='reuser', metavar='USER',
        help='use USER for connections to HOST')

	gpars = pars.add_argument_group('gpg/ssl arguments')
	gpars.add_argument(
        '-r', '--recipients',
        dest='rcp', metavar='ID(s)',
        help='gpg-key ID(s) to use for ' \
             'encryption (string seperated by spaces)')
	gpars.add_argument(
        '-u', '--user',
        dest='usr', metavar='USER', default=cfgs['user'],
        help='query entrys only for USER ' \
             '(defaults to current user, overridden by -A)')
	gpars.add_argument(
        '-x', '--x509',
        dest='gpv', action='store_const', const='gpgsm',
        help='force ssl compatible gpgsm mode - usually is autodetected ' \
             '(use --cert --key for imports)')
	gpars.add_argument(
        '-C', '--cert',
        dest='sslcrt', metavar='SSL-Certificate',
        help='one-shot setting of SSL-Certificate')
	gpars.add_argument(
        '-K', '--key',
        dest='sslkey', metavar='SSL-Private-Key',
        help='one-shot setting of SSL-Private-Key')
	gpars.add_argument(
        '--ca', '--ca-cert',
        dest='sslca', metavar='SSL-CA-Certificate',
        help='one-shot setting of SSL-CA-Certificate')
	gpars.add_argument(
        '-P', '--passcrypt',
        dest='pcr', metavar='CRYPTFILE',
        default=path.expanduser('~/.passcrypt'),
        help='set location of CRYPTFILE to use for gpg features')
	gpars.add_argument(
        '-Y', '--yaml',
        dest='yml', metavar='YAMLFILE',
        default=path.expanduser('~/.pwd.yaml'),
        help='set location of one-time password YAMLFILE to read & delete')
	gpars.add_argument(
        '-S', '--slot',
        dest='ysl', default=None, type=int, choices=(1, 2),
        help='set one of the two slots on the yubi-key (only useful for -y)')

	ypars = pars.add_argument_group('yubikey arguments')
	ypars.add_argument(
        '-y', '--ykserial',
        nargs='?', dest='yks', metavar='SERIAL', default=False,
        help='switch to yubikey mode and optionally set SERIAL of yubikey')

	gpars = pars.add_argument_group('action arguments')
	gpars.add_argument(
        '-a', '--add',
        dest='add', metavar='ENTRY',
        help='add ENTRY (password will be asked interactivly)')
	gpars.add_argument(
        '-c', '--change',
        dest='chg', metavar='ENTRY',
        help='change ENTRY (password will be asked interactivly)')
	gpars.add_argument(
        '-d', '--delete',
        dest='rms', metavar='ENTRY', nargs='+',
        help='delete ENTRY(s) from the passcrypt list')
	gpars.add_argument(
        '-l', '--list',
        nargs='?', dest='lst', metavar='PATTERN', default=False,
        help='search entry matching PATTERN if given otherwise list all')

	autocomplete(pars)
	args = pars.parse_args()
	if args.hlp or (args.yks is False and args.lst is False and \
	      args.add is None and args.chg is None and \
	     args.rms is None and (args.sslcrt is None and args.sslkey is None)):
		pars.print_help()
		exit(1)
	return pars, args

def gui(typ='pw'):
	"""gui wrapper function to not run unnecessary code"""
	poclp, boclp = paste('pb')
	pars, cfgs = confpars()
	if args.yks is False and args.lst is False and \
	      args.add is None and args.chg is None and \
	     args.rms is None and (args.sslcrt is None and args.sslkey is None):
		pars.print_help()
	exit()
	if typ == 'yk':
		__in = xgetpass()
		__res = ykchalres(__in, cfgs['ykslot'], cfgs['ykser'])
		if not __res:
			xmsgok('no entry found for %s or decryption failed'%__in)
			exit(1)
		forkwaitclip(__res, poclp, boclp, cfgs['time'])
	pcm = PassCrypt(**cfgs)
	__in = xgetpass()
	if not __in: exit(1)
	__ent = pcm.lspw(__in)
	if __ent and __in:
		if __in not in __ent.keys() or not __ent[__in]:
			xmsgok('no entry found for %s'%__in)
			exit(1)
		__pc = __ent[__in]
		if __pc:
			if len(__pc) == 2:
				xnotify('%s: %s'%(__in, __pc[1]), cfgs['time'])
			poclp, boclp = paste('pb')
			forkwaitclip(__pc[0], poclp, boclp, cfgs['time'])


if __name__ == '__main__':
	exit(1)
